ssstar/src/objstore/s3.rs

```diff
@@ -1468,7 +1474,125 @@ async fn make_s3_client(
+    let s3_config = s3_config_builder.build();
+
+    let smithy_client = make_smithy_client(config, &s3_config);
+
+    Ok(aws_sdk_s3::Client::with_config(smithy_client, s3_config))
+}
+
+/// Make an AWS SDK [`aws_smithy_client::erase::DynClient`] instance that is pre-configured with middleware
+/// that will set the user agent to the value specified in `config`, if any.
+///
+/// This ugly bit of hackery is mostly copy-pasted from
+/// [`aws_sdk_s3::Client::from_conf`](https://docs.rs/aws-sdk-s3/0.24.0/aws_sdk_s3/struct.Client.html#method.from_conf),
+/// but modified to add some custom middleware that will overwrite whatever user agent string the
+/// AWS SDK normally uses, replacing it with a user-provided user agent from our [`Config`] type.
+fn make_smithy_client(
+    ssstar_config: &Config,
+    s3_config: &aws_sdk_s3::Config,
+) -> aws_smithy_client::erase::DynClient<aws_smithy_client::retry::Standard> {
+    // Make a custom Smithy client so that we can fully control the user agent.
+    //
+    // Inspired by the approach used in https://github.com/awslabs/amazon-qldb-shell/blob/main/src/awssdk_driver.rs
+
+    /// A [`aws_smithy_http::middleware::MapRequest`] impl which modifies Smithy HTTP requests to
+    /// replace whatever User-Agent header was present, if any, with a specific user agent string.
+    #[derive(Clone, Debug)]
+    struct UserAgent(Option<String>);
+
+    impl MapRequest for UserAgent {
+        type Error = Box<dyn Error + Send + Sync + 'static>;
+
+        fn name(&self) -> &'static str {
+            stringify!(UserAgent)
+        }
+
+        fn apply(&self, request: operation::Request) -> Result<operation::Request, Self::Error> {
+            if let Some(ua) = self.0.as_ref() {
+                request.augment(|mut req, _conf| {
+                    req.headers_mut().insert(
+                        HeaderName::from_static("user-agent"),
+                        HeaderValue::from_str(ua)?,
+                    );
+                    Ok(req)
+                })
+            } else {
+                Ok(request)
+            }
+        }
+    }
+
+    // ********************
+    // this is where the copy-paste from aws_sdk_s3::Client::from_conf starts.  that code uses the
+    // var name `conf` for the S3 client config, and to avoid too many changes in the event we have
+    // to copy-paste this again from an updated version, we'll do the same here.
+    let conf = s3_config;
+
+    let retry_config = conf
+        .retry_config()
+        .cloned()
+        .unwrap_or_else(aws_smithy_types::retry::RetryConfig::disabled);
+    let timeout_config = conf
+        .timeout_config()
+        .cloned()
+        .unwrap_or_else(aws_smithy_types::timeout::TimeoutConfig::disabled);
+    let sleep_impl = conf.sleep_impl();
+    if (retry_config.has_retry() || timeout_config.has_timeouts()) && sleep_impl.is_none() {
+        panic!("An async sleep implementation is required for retries or timeouts to work. \
```

Comment:
It's not ideal to panic in library crate, but if this should never happen then it may be ok

Reply:
I agree but this is copied from the AWS SDK, so we're already calling this code that can panic when we init the SDK.  I've just made it more explicit )